### ISSUE TITLE
unexpected panic with `cargo check`

### ISSUE THREAD

**Issue by nloadholtes** [rust-lang/rust#105726]
### Code

Hi, I am still really new to rust and tried to make a minimal version of this, but kept running into issues. My apologies if this makes the report harder to work with.

I'm working on learning rust by contributing to rust-libp2. [Here's my PR](https://github.com/libp2p/rust-libp2p/pull/3224). The very last commit is the one that generated the backtrace below. I'm using VSCode as my IDE and it did not report any errors.

Please let me know if there's more steps anyone wants me to do, I've never crashed a compiler before so I'm pretty excited about this and want to help out. :smile: 

```Rust
// This is the last line I changed before the error occurred,
// I basically added in the `#[from]` then ran `cargo check --package=libp2p-floodsub`
pub struct CodecError(#[from] prost_codec::Error);
```

### Meta

`rustc --version --verbose`:
```
rustc 1.65.0 (897e37553 2022-11-02)
binary: rustc
commit-hash: 897e37553bba8b42751c67658967889d11ecd120
commit-date: 2022-11-02
host: x86_64-unknown-linux-gnu
release: 1.65.0
LLVM version: 15.0.0

```

### Error output

```
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.65.0 (897e37553 2022-11-02) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type lib -C embed-bitcode=no -C debuginfo=2 -C incremental=[REDACTED]

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [evaluate_obligation] evaluating trait selection obligation `protocol::FloodsubProtocol: libp2p_swarm::upgrade::InboundUpgradeSend`
#1 [check_well_formed] checking that `layer::Floodsub` is well-formed
#2 [analysis] running analysis passes on this crate
end of query stack

```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
stack backtrace:
   0:     0x7f9a8304fd40 - std::backtrace_rs::backtrace::libunwind::trace::h32eb3e08e874dd27
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7f9a8304fd40 - std::backtrace_rs::backtrace::trace_unsynchronized::haa3f451d27bc11a5
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f9a8304fd40 - std::sys_common::backtrace::_print_fmt::h5b94a01bb4289bb5
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x7f9a8304fd40 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hb070b7fa7e3175df
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x7f9a830aabfe - core::fmt::write::hd5207aebbb9a86e9
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/fmt/mod.rs:1202:17
   5:     0x7f9a83040935 - std::io::Write::write_fmt::h3bd699bbd129ab8a
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/io/mod.rs:1679:15
   6:     0x7f9a830529f3 - std::sys_common::backtrace::_print::h7a21be552fdf58da
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x7f9a830529f3 - std::sys_common::backtrace::print::ha85c41fe4dd80b13
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x7f9a830529f3 - std::panicking::default_hook::{{closure}}::h04cca40023d0eeca
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:295:22
   9:     0x7f9a830526df - std::panicking::default_hook::haa3ca8c310ed5402
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:314:9
  10:     0x7f9a81a452e1 - rustc_driver[cfb34b1539811fe8]::DEFAULT_HOOK::{closure#0}::{closure#0}
  11:     0x7f9a8305322d - std::panicking::rust_panic_with_hook::h7b190ce1a948faac
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:702:17
  12:     0x7f9a83053087 - std::panicking::begin_panic_handler::{{closure}}::hbafbfdc3e1b97f68
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:588:13
  13:     0x7f9a830501ec - std::sys_common::backtrace::__rust_end_short_backtrace::hda93e5fef243b4c0
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/sys_common/backtrace.rs:138:18
  14:     0x7f9a83052da2 - rust_begin_unwind
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:584:5
  15:     0x7f9a830a77d3 - core::panicking::panic_fmt::h8d17ca1073d9a733
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:142:14
  16:     0x7f9a82a62f61 - <rustc_query_system[860ed7b39cdfb46b]::dep_graph::dep_node::DepNode<rustc_middle[a9ca1c3f9fd197cd]::dep_graph::dep_node::DepKind> as rustc_middle[a9ca1c3f9fd197cd]::dep_graph::dep_node::DepNodeExt>::extract_def_id::{closure#0}
  17:     0x7f9a80379e3b - <rustc_query_system[860ed7b39cdfb46b]::dep_graph::dep_node::DepNode<rustc_middle[a9ca1c3f9fd197cd]::dep_graph::dep_node::DepKind> as rustc_middle[a9ca1c3f9fd197cd]::dep_graph::dep_node::DepNodeExt>::extract_def_id
  18:     0x7f9a813616bc - rustc_query_impl[d6191eea714bd250]::plumbing::force_from_dep_node::<rustc_query_impl[d6191eea714bd250]::queries::local_def_id_to_hir_id>
  19:     0x7f9a804cb291 - <rustc_middle[a9ca1c3f9fd197cd]::ty::context::TyCtxt as rustc_query_system[860ed7b39cdfb46b]::dep_graph::DepContext>::try_force_from_dep_node
  20:     0x7f9a804ca022 - <rustc_query_system[860ed7b39cdfb46b]::dep_graph::graph::DepGraph<rustc_middle[a9ca1c3f9fd197cd]::dep_graph::dep_node::DepKind>>::try_mark_previous_green::<rustc_query_impl[d6191eea714bd250]::plumbing::QueryCtxt>
  21:     0x7f9a804c9ff0 - <rustc_query_system[860ed7b39cdfb46b]::dep_graph::graph::DepGraph<rustc_middle[a9ca1c3f9fd197cd]::dep_graph::dep_node::DepKind>>::try_mark_previous_green::<rustc_query_impl[d6191eea714bd250]::plumbing::QueryCtxt>
  22:     0x7f9a804c9ff0 - <rustc_query_system[860ed7b39cdfb46b]::dep_graph::graph::DepGraph<rustc_middle[a9ca1c3f9fd197cd]::dep_graph::dep_node::DepKind>>::try_mark_previous_green::<rustc_query_impl[d6191eea714bd250]::plumbing::QueryCtxt>
  23:     0x7f9a804c9ff0 - <rustc_query_system[860ed7b39cdfb46b]::dep_graph::graph::DepGraph<rustc_middle[a9ca1c3f9fd197cd]::dep_graph::dep_node::DepKind>>::try_mark_previous_green::<rustc_query_impl[d6191eea714bd250]::plumbing::QueryCtxt>
  24:     0x7f9a80f1f0ca - rustc_query_system[860ed7b39cdfb46b]::query::plumbing::get_query::<rustc_query_impl[d6191eea714bd250]::queries::evaluate_obligation, rustc_query_impl[d6191eea714bd250]::plumbing::QueryCtxt>
  25:     0x7f9a80f1e235 - <rustc_query_impl[d6191eea714bd250]::Queries as rustc_middle[a9ca1c3f9fd197cd]::ty::query::QueryEngine>::evaluate_obligation
  26:     0x7f9a806dd937 - <rustc_trait_selection[f6cb7b5468c1337d]::traits::fulfill::FulfillProcessor as rustc_data_structures[62ce3989ce8140ee]::obligation_forest::ObligationProcessor>::process_obligation
  27:     0x7f9a806dad0f - <rustc_data_structures[62ce3989ce8140ee]::obligation_forest::ObligationForest<rustc_trait_selection[f6cb7b5468c1337d]::traits::fulfill::PendingPredicateObligation>>::process_obligations::<rustc_trait_selection[f6cb7b5468c1337d]::traits::fulfill::FulfillProcessor, rustc_data_structures[62ce3989ce8140ee]::obligation_forest::Outcome<rustc_trait_selection[f6cb7b5468c1337d]::traits::fulfill::PendingPredicateObligation, rustc_infer[ed26b14e1208c12f]::traits::FulfillmentErrorCode>>
  28:     0x7f9a80e2e5b0 - <rustc_trait_selection[f6cb7b5468c1337d]::traits::engine::ObligationCtxt>::select_all_or_error
  29:     0x7f9a811e92d3 - <rustc_infer[ed26b14e1208c12f]::infer::InferCtxtBuilder>::enter::<(), rustc_typeck[4a3b2f91c49b3c3e]::check::wfcheck::enter_wf_checking_ctxt<rustc_typeck[4a3b2f91c49b3c3e]::check::wfcheck::check_type_defn<rustc_typeck[4a3b2f91c49b3c3e]::check::wfcheck::check_item::{closure#2}>::{closure#0}>::{closure#0}>
  30:     0x7f9a811c5ce4 - rustc_typeck[4a3b2f91c49b3c3e]::check::wfcheck::check_well_formed
  31:     0x7f9a80afa231 - <rustc_query_system[860ed7b39cdfb46b]::dep_graph::graph::DepGraph<rustc_middle[a9ca1c3f9fd197cd]::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle[a9ca1c3f9fd197cd]::ty::context::TyCtxt, rustc_span[8f00505efff112e2]::def_id::LocalDefId, ()>
  32:     0x7f9a80af8cb7 - rustc_query_system[860ed7b39cdfb46b]::query::plumbing::try_execute_query::<rustc_query_impl[d6191eea714bd250]::plumbing::QueryCtxt, rustc_query_system[860ed7b39cdfb46b]::query::caches::DefaultCache<rustc_span[8f00505efff112e2]::def_id::LocalDefId, ()>>
  33:     0x7f9a81775822 - rustc_query_system[860ed7b39cdfb46b]::query::plumbing::force_query::<rustc_query_impl[d6191eea714bd250]::queries::check_well_formed, rustc_query_impl[d6191eea714bd250]::plumbing::QueryCtxt>
  34:     0x7f9a81775690 - rustc_query_impl[d6191eea714bd250]::plumbing::force_from_dep_node::<rustc_query_impl[d6191eea714bd250]::queries::check_well_formed>
  35:     0x7f9a804cb291 - <rustc_middle[a9ca1c3f9fd197cd]::ty::context::TyCtxt as rustc_query_system[860ed7b39cdfb46b]::dep_graph::DepContext>::try_force_from_dep_node
  36:     0x7f9a804ca022 - <rustc_query_system[860ed7b39cdfb46b]::dep_graph::graph::DepGraph<rustc_middle[a9ca1c3f9fd197cd]::dep_graph::dep_node::DepKind>>::try_mark_previous_green::<rustc_query_impl[d6191eea714bd250]::plumbing::QueryCtxt>
  37:     0x7f9a80af8352 - rustc_query_system[860ed7b39cdfb46b]::query::plumbing::ensure_must_run::<rustc_query_impl[d6191eea714bd250]::plumbing::QueryCtxt, rustc_span[8f00505efff112e2]::def_id::LocalDefId, ()>
  38:     0x7f9a814aab75 - rustc_query_system[860ed7b39cdfb46b]::query::plumbing::get_query::<rustc_query_impl[d6191eea714bd250]::queries::check_mod_type_wf, rustc_query_impl[d6191eea714bd250]::plumbing::QueryCtxt>
  39:     0x7f9a80dbb358 - rustc_data_structures[62ce3989ce8140ee]::sync::par_for_each_in::<&[rustc_span[8f00505efff112e2]::def_id::LocalDefId], <rustc_middle[a9ca1c3f9fd197cd]::hir::map::Map>::par_for_each_module<rustc_typeck[4a3b2f91c49b3c3e]::check_crate::{closure#5}::{closure#0}::{closure#0}>::{closure#0}>
  40:     0x7f9a80dbb163 - <rustc_session[c5946fac61f8bc34]::session::Session>::track_errors::<rustc_typeck[4a3b2f91c49b3c3e]::check_crate::{closure#5}, ()>
  41:     0x7f9a80db85e0 - rustc_typeck[4a3b2f91c49b3c3e]::check_crate
  42:     0x7f9a80db7c27 - rustc_interface[3182dd864eff9d7d]::passes::analysis
  43:     0x7f9a81816382 - <rustc_query_system[860ed7b39cdfb46b]::dep_graph::graph::DepGraph<rustc_middle[a9ca1c3f9fd197cd]::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle[a9ca1c3f9fd197cd]::ty::context::TyCtxt, (), core[8c92e53db3fc2eaa]::result::Result<(), rustc_errors[f77a66b68db622d5]::ErrorGuaranteed>>
  44:     0x7f9a81815a4a - rustc_query_system[860ed7b39cdfb46b]::query::plumbing::try_execute_query::<rustc_query_impl[d6191eea714bd250]::plumbing::QueryCtxt, rustc_query_system[860ed7b39cdfb46b]::query::caches::DefaultCache<(), core[8c92e53db3fc2eaa]::result::Result<(), rustc_errors[f77a66b68db622d5]::ErrorGuaranteed>>>
  45:     0x7f9a81815517 - rustc_query_system[860ed7b39cdfb46b]::query::plumbing::get_query::<rustc_query_impl[d6191eea714bd250]::queries::analysis, rustc_query_impl[d6191eea714bd250]::plumbing::QueryCtxt>
  46:     0x7f9a80901897 - <rustc_interface[3182dd864eff9d7d]::passes::QueryContext>::enter::<rustc_driver[cfb34b1539811fe8]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[8c92e53db3fc2eaa]::result::Result<(), rustc_errors[f77a66b68db622d5]::ErrorGuaranteed>>
  47:     0x7f9a808f528c - rustc_interface[3182dd864eff9d7d]::interface::create_compiler_and_run::<core[8c92e53db3fc2eaa]::result::Result<(), rustc_errors[f77a66b68db622d5]::ErrorGuaranteed>, rustc_driver[cfb34b1539811fe8]::run_compiler::{closure#1}>
  48:     0x7f9a808f3ac1 - <scoped_tls[e395fa6e23b19669]::ScopedKey<rustc_span[8f00505efff112e2]::SessionGlobals>>::set::<rustc_interface[3182dd864eff9d7d]::interface::run_compiler<core[8c92e53db3fc2eaa]::result::Result<(), rustc_errors[f77a66b68db622d5]::ErrorGuaranteed>, rustc_driver[cfb34b1539811fe8]::run_compiler::{closure#1}>::{closure#0}, core[8c92e53db3fc2eaa]::result::Result<(), rustc_errors[f77a66b68db622d5]::ErrorGuaranteed>>
  49:     0x7f9a808f37af - std[71cb4861428b0c25]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[3182dd864eff9d7d]::util::run_in_thread_pool_with_globals<rustc_interface[3182dd864eff9d7d]::interface::run_compiler<core[8c92e53db3fc2eaa]::result::Result<(), rustc_errors[f77a66b68db622d5]::ErrorGuaranteed>, rustc_driver[cfb34b1539811fe8]::run_compiler::{closure#1}>::{closure#0}, core[8c92e53db3fc2eaa]::result::Result<(), rustc_errors[f77a66b68db622d5]::ErrorGuaranteed>>::{closure#0}, core[8c92e53db3fc2eaa]::result::Result<(), rustc_errors[f77a66b68db622d5]::ErrorGuaranteed>>
  50:     0x7f9a818a8d70 - <<std[71cb4861428b0c25]::thread::Builder>::spawn_unchecked_<rustc_interface[3182dd864eff9d7d]::util::run_in_thread_pool_with_globals<rustc_interface[3182dd864eff9d7d]::interface::run_compiler<core[8c92e53db3fc2eaa]::result::Result<(), rustc_errors[f77a66b68db622d5]::ErrorGuaranteed>, rustc_driver[cfb34b1539811fe8]::run_compiler::{closure#1}>::{closure#0}, core[8c92e53db3fc2eaa]::result::Result<(), rustc_errors[f77a66b68db622d5]::ErrorGuaranteed>>::{closure#0}, core[8c92e53db3fc2eaa]::result::Result<(), rustc_errors[f77a66b68db622d5]::ErrorGuaranteed>>::{closure#1} as core[8c92e53db3fc2eaa]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  51:     0x7f9a8305d003 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h49f797984e2121bf
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/alloc/src/boxed.rs:1940:9
  52:     0x7f9a8305d003 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hfa4f3d0ee6440e0b
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/alloc/src/boxed.rs:1940:9
  53:     0x7f9a8305d003 - std::sys::unix::thread::Thread::new::thread_start::h62ca48b42d48a8fc
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/sys/unix/thread.rs:108:17
  54:     0x7f9a7ec836db - start_thread
                               at /build/glibc-CVJwZb/glibc-2.27/nptl/pthread_create.c:463
  55:     0x7f9a7e9ac61f - clone
                               at /build/glibc-CVJwZb/glibc-2.27/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:95
  56:                0x0 - <unknown>


```

</p>
</details>

**Comment by nloadholtes** [comment]
Hmmm, a friend suggested running `cargo clean` and then rechecking. I did this and the ICE did not occur again. I'm going to close this and assume it was a fluke.
